File: 06. Backtracking/01. Subsets/main.py
def generateSubsets(index, nums, current, result):
    if index == len(nums):
        result.append(current[:])  # Store a copy of current subset
        return

    # Exclude the current element
    generateSubsets(index + 1, nums, current, result)

    # Include the current element
    current.append(nums[index])
    generateSubsets(index + 1, nums, current, result)
    current.pop()  # Backtrack

def subsets(nums):
    result = []
    generateSubsets(0, nums, [], result)
    return result

def main():
    nums = [1, 2, 3]
    print("Subsets using Backtracking:")
    print(subsets(nums))

if __name__ == "__main__":
    main()

File: 06. Backtracking/01. Subsets/main.js
function generateSubsets(index, nums, current, result) {
    if (index === nums.length) {
        result.push([...current]);
        return;
    }

    // Exclude the current element
    generateSubsets(index + 1, nums, current, result);

    // Include the current element
    current.push(nums[index]);
    generateSubsets(index + 1, nums, current, result);
    current.pop(); // Backtrack
}

function subsets(nums) {
    let result = [];
    generateSubsets(0, nums, [], result);
    return result;
}


function main() {
    let nums = [1, 2, 3];
    console.log("Subsets using Backtracking:");
    console.log(subsets(nums));    
}

main();